akka-stream/src/main/scala/akka/stream/impl/BoundedSourceQueue.scala

```diff
@@ -101,6 +108,13 @@ import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, Out
           case Done(QueueOfferResult.Failure(ex)) => failStage(ex)
           case NeedsActivation                    => throw new IllegalStateException // needs to be cleared before
         }
+
+      def notifySubscribersIfNeeded(): Unit = {
+        // if the queue was full previously but now it can accept a new element
+        if (readyToAccept.compareAndSet(false, true)) {
+          subscriptions.forEach(promise => promise.success(akka.Done.getInstance()))
```

Comment:
Here we notify **all** subscribers. Usually there should not be many (otherwise someone uses this `whenReady` improperly). We can think about notifying only a part of subscribers, but I don't feel this logic is really required. What do you think?

Reply:
Good catch! I've fixed that and added a separate test to ensure subscriptions are being cleaned up properly.